Source/ProgressionSystemRuntime/Private/Widgets/PSOverlayWidget.cpp

```diff
@@ -4,11 +4,113 @@
-void UPSOverlayWidget::SetOverlayVisibility(ESlateVisibility VisibilitySlate)
+// Sets the visibility of the overlay elements and playing fade animation if needed
+void UPSOverlayWidget::SetOverlayVisibility(ESlateVisibility VisibilitySlate, bool bShouldPlayFadeAnimation/* = false*/)
+{
+	if (!bShouldPlayFadeAnimation)
+	{
+		SetOverlayItemsVisibility(VisibilitySlate);
+		return;
+	}
+
+	bShouldPlayFadeAnimationInternal = bShouldPlayFadeAnimation;
+
+	if (VisibilitySlate == ESlateVisibility::Visible)
+	{
+		ESlateVisibility PrevOverlayOpacity = PSCBackgroundOverlay->GetVisibility();
+		ESlateVisibility PrevIconOpacity = PSCBackgroundIconLock->GetVisibility();
+
+		//if new visibility is same as previous animation is not required 
+		if (PrevOverlayOpacity == VisibilitySlate && PrevIconOpacity == VisibilitySlate)
+		{
+			bShouldPlayFadeAnimationInternal = false;
+		}
+		bIsFadeInAnimationInternal = true;
+		SetOverlayItemsVisibility(VisibilitySlate);
+	}
+	else
+	{
+		bIsFadeInAnimationInternal = false;
+	}
+
+	const UWorld* World = GetWorld();
+	if (!ensureMsgf(World, TEXT("ASSERT: [%i] %s:\n'World' is not valid!"), __LINE__, *FString(__FUNCTION__)))
+	{
+		return;
+	}
+
+	StartTimeFadeAnimationInternal = World->GetTimeSeconds();
+}
+
+// overrides NativeTick to make the user widget tickable
+void UPSOverlayWidget::NativeTick(const FGeometry& MyGeometry, float InDeltaTime)
+{
+	Super::NativeTick(MyGeometry, InDeltaTime);
+	if (bShouldPlayFadeAnimationInternal)
+	{
+		const bool bIsFinished = !FadeOverlayElementsAnimation(StartTimeFadeAnimationInternal);
+		if (bIsFinished)
+		{
+			bShouldPlayFadeAnimationInternal = false;
+			StartTimeFadeAnimationInternal = 0.f;
+			if (!bIsFadeInAnimationInternal)
+			{
+				SetOverlayItemsVisibility(ESlateVisibility::Collapsed);
+			}
+		}
+	}
+}
+
+// Event to execute when widget is ready
+void UPSOverlayWidget::NativeConstruct()
+{
+	Super::NativeConstruct();
+	FadeCurveFloatInternal = UPSDataAsset::Get().GetFadeCurveFloat();
+}
+
+// Play the overlay elemetns fade-in/fade-out animation. Uses the internal FadeCurveFloatInternal initialized in NativeConstruct
+bool UPSOverlayWidget::FadeOverlayElementsAnimation(float& StartTimeRef)
+{
+	if (!StartTimeRef || !FadeCurveFloatInternal || !bShouldPlayFadeAnimationInternal)
+	{
+		StartTimeRef = 0.f;
+		return false;
+	}
+
+	const UWorld* World = GetWorld();
+	check(World);
+
+	const float SecondsSinceStart = GetWorld()->GetTimeSeconds() - StartTimeRef;
+	float OpacityValue = FadeCurveFloatInternal->GetFloatValue(SecondsSinceStart);
+
+	float MinTime = 0.f;
+	float MaxTime = 0.f;
+	FadeCurveFloatInternal->GetTimeRange(MinTime, MaxTime);
+	if (SecondsSinceStart >= MaxTime)
+	{
+		// The curve is finished
+		return false;
+	}
+
+	if (!bIsFadeInAnimationInternal)
+	{
+		OpacityValue = MaxTime - OpacityValue;
+	}
+
+	if (PSCBackgroundOverlay && PSCBackgroundIconLock)
+	{
+		PSCBackgroundOverlay->SetRenderOpacity(OpacityValue);
```

Comment:
There are a few problems:

- Canvas is expensive. We should never use it when other canvases are already spawned (we already have one in Menu), as it causes performance issues and bugs. Instead, build the hierarchy from subwidgets with offsets and alignments to achieve the desired result. Instead of a canvas, use UE's `Overlay` widget as the root of all other subwidgets, and place both images inside it. For the lock image, you will likely want to additionally wrap it in a `SizeBox` with offsets and alignments to locate in the center.

- You are calling `SetRenderOpacity` twice, exposing artistic subwidgets to code. These subwidgets might change, be removed, or have more icons added. We typically only expose the root element, like UE's Overlay, to the code. This way, when you change the opacity of the main subwidget, the rest will change as well. This approach gives UI designers the freedom to use whatever elements they want. Please remove both subwidgets from the code and expose the root element instead.

Reply:
done: https://github.com/h4rdmol/ProgressionSystem/pull/20/commits/c099e10b0f70a91783e380ee42eaeb5c1867aef3

Reply:
![image](https://github.com/user-attachments/assets/577d6b35-e3b6-4a65-83a2-e9b4a3c72f6b)
![image](https://github.com/user-attachments/assets/db577b7c-b84f-4d79-a35e-eb41ae2f5722)
![image](https://github.com/user-attachments/assets/abf77842-7bee-4ba6-ab58-6f6d8d488844)
![image](https://github.com/user-attachments/assets/77de92cb-a3a7-4be7-a85d-c2b95417085a)